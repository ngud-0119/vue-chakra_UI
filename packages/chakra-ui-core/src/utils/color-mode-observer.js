import Vue from 'vue'

export const defineColorModeObserver = ({ colorMode, theme, icons, toggleColorMode } = {}) => Vue.observable({
  colorMode,
  theme,
  icons,
  toggleColorMode
})

/**
 * This observed store object observed the colorMode and stores it in an
 * observed object that other components can consume.
 */
export const colorModeObserver = defineColorModeObserver({
  toggleColorMode: emitToggleColorMode
})

/**
 * Utility function that returns a value based on the colorMode
 * @param {string | number | Array<string | number>} lightValue Value when colorMode is `light`
 * @param {string | number | Array<string | number>} darkValue Value when colorMode is `dark`
 * @param {import('Vue').ComponentOptions<Vue, { colorMode: 'light' | 'dark', theme: any, icons: any }>} observer Value when colorMode is `dark`
 * @return {string | number | Array<string | number>}
 */
export const mode = (lightValue, darkValue, observer) => {
  const { colorMode } = observer || colorModeObserver
  return colorMode === 'dark' ? darkValue : lightValue
}

const changeColorModeListeners = []

export const colorModeObserverEventBus = new Vue()

colorModeObserverEventBus.$on('change:colorMode', (newVal) => {
  colorModeObserver.colorMode = newVal
  changeColorModeListeners.forEach(handler => handler(newVal))
})

export function emitToggleColorMode () {
  colorModeObserverEventBus.$emit('command:toggleColorMode')
}

export const onUpdateColorMode = fn => changeColorModeListeners.push(fn)
