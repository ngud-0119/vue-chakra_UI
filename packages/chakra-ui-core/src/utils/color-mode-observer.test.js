import { mode, defineColorModeObserver } from './color-mode-observer'

describe('colorModeObserver', () => {
  it('`colorModeObserver` should provide and observe values', () => {
    const baseTheme = {
      colors: {
        success: '#3ea76a',
        warning: '#ffc61a',
        error: '#dd3b4b'
      },
      shadows: {
        outline: '0 0 0 3px rgba(66, 153, 225, 0.6)',
        inner: 'inset 0 2px 4px 0 rgba( 0, 0, 0, 0.06)'
      },
      baseStyles: {
        CButton: {
          bg: 'success',
          borderRadius: '200px',
          shadow: 'inner'
        }
      }
    }

    const observer = defineColorModeObserver()

    observer.theme = baseTheme
    observer.colorMode = 'light'
    expect(observer.theme).toEqual(baseTheme)
    expect(observer.colorMode).toEqual('light')

    // Mutate `colorMode`
    observer.colorMode = 'dark'
    const readColorMode = observer.colorMode
    expect(readColorMode).toBe('dark')
    expect(readColorMode).toEqual(observer.colorMode)

    // Mutate `theme`
    const newTheme = Object.assign(baseTheme, {
      colors: {
        info: '#0000FF'
      }
    })
    observer.theme = newTheme
    const readTheme = observer.theme
    expect(readTheme).toMatchObject(newTheme)
    expect(readTheme).toMatchSnapshot()
  })

  it('`mode` function should return current colorMode value', () => {
    const observer = defineColorModeObserver({
      colorMode: 'dark'
    })

    const bg = mode('red', 'blue', observer)
    expect(bg).toBe('blue')
    observer.colorMode = 'light'

    // We invoke as a function since we are in a non-reactive context
    const _bg = mode('red', 'blue', observer)
    expect(_bg).toBe('red')
  })
})
