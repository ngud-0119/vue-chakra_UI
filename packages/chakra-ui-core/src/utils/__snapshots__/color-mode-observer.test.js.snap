// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colorModeObserver \`colorModeObserver\` should provide and observe values 1`] = `
Object {
  "baseStyles": Object {
    "CButton": Object {
      "bg": "success",
      "borderRadius": "200px",
      "shadow": "inner",
    },
  },
  "colors": Object {
    "info": "#0000FF",
  },
  "shadows": Object {
    "inner": "inset 0 2px 4px 0 rgba( 0, 0, 0, 0.06)",
    "outline": "0 0 0 3px rgba(66, 153, 225, 0.6)",
  },
}
`;
